### PR TITLE
persist: call Indexed::step less frequently.

### PR DESCRIPTION
**Pull request by ruchirK**
This commit teaches the persistence runtime to try to grab as many commands as possible
execute them in a sequence, and then call Indexed::step rather than calling it after
every single write and read. Doing so should amortize the cost of creating Future and
Trace batches over more commands.

### CODE REVIEW

**Comment by ruchirK** [comment]
@danhhz this was my understanding of the commit to batch up step that we talked about offline. This commit seems a lot more performant on my laptop than the previous commit that tried to perform a step once ever 100 ms - everything seems to respond with < 200 ms latency when under light load + persistent system tables.

It's still a little bit flaky with nemesis tests -- `cargo-stress` is extremely useful for sussing this out and I'm working on understanding why. I think it has to do with when a `StorageAvailable` nemesis command is interleaved between two successive requests to indexed that get processed in the same invocation of `work`

File: src/persist/src/indexed/runtime.rs [hunk 1]
```diff
@@ -578,52 +578,67 @@ impl<U: Buffer, L: Blob> RuntimeImpl<U, L> {
...
+                    let r = self.indexed.snapshot(id);
+                    res.fill(r);
+                }
+                Cmd::Listen(id, listen_fn, res) => {
+                    let r = self.indexed.listen(id, listen_fn);
+                    res.fill(r);
+                }
             }
         }
+        let _ = self.indexed.step();
```
Comment (ruchirK) [inline src/persist/src/indexed/runtime.rs:641]: I couldn't think of a good way to pass this response back (which command do you send it to?) or a good reason to (even if step happened to fail it doesn't really affect the success or failure of the other commands)

File: src/persist/src/indexed/runtime.rs [hunk 2]
```diff
@@ -578,52 +578,67 @@ impl<U: Buffer, L: Blob> RuntimeImpl<U, L> {
...
-            }
-            Cmd::AllowCompaction(id, ts, res) => {
-                let r = self.indexed.allow_compaction(id, ts);
-                res.fill(r);
-            }
-            Cmd::Snapshot(id, res) => {
-                let r = self.indexed.snapshot(id);
-                res.fill(r);
+
+        loop {
```
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:591]: toss in a comment on what this is doing for future readers

File: src/persist/src/indexed/runtime.rs [hunk 3]
```diff
@@ -578,52 +578,67 @@ impl<U: Buffer, L: Blob> RuntimeImpl<U, L> {
...
+            match self.rx.try_recv() {
+                Ok(cmd) => cmds.push(cmd),
+                Err(crossbeam_channel::TryRecvError::Empty) => {
+                    break;
+                }
+                Err(crossbeam_channel::TryRecvError::Disconnected) => {
```
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:597]: do we want have this set a bool and return after processing the commands it did read? i'm a slight lean toward yes

File: src/persist/src/indexed/runtime.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/src/indexed/runtime.rs:641]: yep makes sense. lets log::warn! here. it'll be spammy for now but we can fix that later

**Review by danhhz — APPROVED** [review]
LGTM

**Comment by danhhz** [comment]
> I think it has to do with when a StorageAvailable nemesis command is interleaved between two successive requests to indexed that get processed in the same invocation of work

oof yeah. the current impl of validator is cheating in a lot of ways. i've been meaning to add a "sometimes fail" mode to unreliable storage, which would also cause us to think through a bunch of them

File: src/persist/src/indexed/runtime.rs (hunk 1 above)
Comment (ruchirK) [reply src/persist/src/indexed/runtime.rs]: done

File: src/persist/src/indexed/runtime.rs (hunk 3 above)
Comment (ruchirK) [reply src/persist/src/indexed/runtime.rs:605]: done

**Comment by ruchirK** [comment]
I rerequested your review @danhhz because I had to apply some more elbow grease in order to actually get this to pass nemesis tests. The structure in Runtime is unchanged, had to do some error hardening in Indexed and fix a bug in how we read from snapshot sometimes.

File: src/persist/src/indexed/mod.rs [hunk 4]
```diff
@@ -454,6 +547,14 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
             }
         }
 
+        for (id, new_future_ts_lower) in future_ts_lower_updates {
```
Comment (ruchirK) [inline src/persist/src/indexed/mod.rs:550]: maybe add a comment here indicating "now that all of the trace writes / metadata writes are committed, its safe to truncate future, and it also doesn't matter if truncating future fails that can be best effort (it doesnt prompt a rollback)"

File: src/persist/src/indexed/mod.rs [hunk 5]
```diff
@@ -131,6 +131,52 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         Ok(indexed)
     }
 
+    /// Revert the in-memory state back to a previously serialized version.
+    ///
+    /// Used to keep the in-memory and durably stored data structures consistent
+    /// in the presence of errors.
+    fn restore(&mut self, meta: BlobMeta) {
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:138]: I feel like there should be a way for this to share most of its code with new, so restore becomes something like `*self = new_indexed_from_meta(self.blob, meta)`. The next_blob_id stuff is an issue, but they don't really have to be incrementing, so we could find the max of any of them and pass that +1 in as a parameter to new_indexed_from_meta as well and use it for all of them? Thoughts?

File: src/persist/src/indexed/mod.rs [hunk 6]
```diff
@@ -197,6 +245,13 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         self.traces
             .entry(id)
             .or_insert_with_key(|id| BlobTrace::new(BlobTraceMeta::new(*id)));
+
+        if let Err(e) = self.blob.set_meta(self.serialize_meta()) {
+            // We were unable to durably commit the register. Revert back to the
+            // previous version of meta.
+            self.restore(prev);
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:252]: worth pulling out a method for this whole `if let Err(e) = self.blob.set_meta(self.serialize_meta()) { ...; self.restore(prev); ...}` bit? we could also put the compactable log TODO in there and keep it from being copied everywhere

File: src/persist/src/indexed/mod.rs [hunk 7]
```diff
@@ -282,6 +344,10 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         let mut updates_by_id: HashMap<Id, Vec<(SeqNo, (Vec<u8>, Vec<u8>), u64, isize)>> =
             HashMap::new();
         let desc = self.buf.snapshot(|seqno, buf| {
+            // Skip updates not in advance of the last sequence number we have observed.
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:347]: elaborate on why, please. they previously made it into the future, but the buffer truncation failed, i think?

File: src/persist/src/indexed/mod.rs [hunk 8]
```diff
@@ -319,19 +386,32 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
...
             // [future.seqno_upper, self.futures_seqno_upper). Use this to make the
             // seqnos of all the future batches line up.
-            debug_assert_eq!(desc.start, self.futures_seqno_upper);
+
+            // We cannot assert strict equality here because a buffer may have
+            // previously failed to truncate.
+            debug_assert!(desc.start <= self.futures_seqno_upper);
             let new_start = future.seqno_upper()[0];
-            debug_assert!(new_start <= desc.start);
+            debug_assert!(new_start <= self.futures_seqno_upper);
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:394]: i'm probably just being dense, why this change?

**Review by danhhz — COMMENTED** [review]
bravo on chasing down this step stuff! i thought this change was going to be much more straightforward. i guess it's nice to catch this early and break up the next PR a bit

File: src/persist/src/indexed/trace.rs [hunk 9]
```diff
@@ -310,9 +310,8 @@ impl Snapshot<Vec<u8>, Vec<u8>> for TraceSnapshot {
     fn read<E: Extend<((Vec<u8>, Vec<u8>), u64, isize)>>(&mut self, buf: &mut E) -> bool {
         if let Some(batch) = self.updates.pop() {
             buf.extend(batch.updates.iter().cloned());
-            return true;
         }
-        false
+        !self.updates.is_empty()
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:314]: I think the semantics of "read returns true if data was read in OR more data will be read in, read returns false if no data was read in AND no further data will be available" make sense / would work. I kept the semantics as currently written because I felt like that was a smaller change overall.

[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:314]: sorry let me be more specific - it was conceptually simpler for me to just stick to the already defined and agreed upon semantics rather than trying to figure out a new semantics and a new impl of those for BufferSnapshot which would i think have to change its state to remember the first time someone called read on it. Neither of those things are the end of the world but its nice to have a already crisp definition of what I'm trying to build when I'm also trying to sort out a bunch of other bugs :)

do you think its ok to address the snapshot semantics change in a followup pr?

[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:314]: Ha! Yeah I have a slightly different take (which is wildly less defensible than yours) where the behavior of trace/future/indexed/decoded snapshot was the contract and the rustdoc on Snapshot::read and the impl of BufferSnapshot were the ones that were wrong. I'm fine leaving this as a followup, but toss a TODO in there please.

Btw, I think a workable impl for BufferSnapshot is
```rust
impl Snapshot<Vec<u8>, Vec<u8>> for BufferSnapshot {
    fn read<E: Extend<((Vec<u8>, Vec<u8>), u64, isize)>>(&mut self, buf: &mut E) -> bool {
        let had_data = self.1.len() > 0;
        buf.extend(self.1.drain(..));
        had_data
    }
}
```

File: src/persist/src/indexed/mod.rs (hunk 5 above)
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:138]: I think thats correct - let me see if I can whip something up. I think you have to be careful to hold onto the existing listeners which might be why I left this this way last night.

Comment (danhhz) [reply src/persist/src/indexed/mod.rs:138]: Ah the listeners too, that's a good point. I'm also fine going with this approach for now given that we might have to change it anyway when we actually start batching stuff (plus at some point we should maybe do minimal rollbacks?)

File: src/persist/src/indexed/trace.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:314]: Added the todo!

File: src/persist/src/indexed/encoding.rs [hunk 10]
```diff
@@ -54,7 +54,7 @@ pub struct BufferEntry {
 /// - None of the ids in graveyard are present in any of the (string, id) tuples
 ///   in id_mapping.
 /// - The same set of ids are present in id_mapping, futures, and traces.
-/// - For each id, the ts_lower in the future is == the ts_upper in the
+/// - For each id, the ts_lower in the future is <= the ts_upper in the
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:57]: done

File: src/persist/src/indexed/encoding.rs [hunk 11]
```diff
@@ -1258,7 +1258,7 @@ mod tests {
         };
         assert_eq!(b.validate(), Err(Error::from("duplicate trace: Id(0)")));
 
-        // Future ts_lower doesn't match trace ts_upper
+        // Future ts_lower in advance of ts_upper
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:1305]: done

File: src/persist/src/indexed/mod.rs (hunk 5 above)
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:141]: left a todo

File: src/persist/src/indexed/mod.rs [hunk 12]
```diff
@@ -131,6 +131,70 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         Ok(indexed)
     }
 
+    /// Revert the in-memory state back to a previously serialized version.
+    ///
+    /// Used to keep the in-memory and durably stored data structures consistent
+    /// in the presence of errors.
+    /// TODO: can we simplify this logic and combine it with Indexed::new()? In
```
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:138]: nit: empty line between paragraphs (there's a couple instances of this)

File: src/persist/src/indexed/mod.rs (hunk 8 above)
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:410]: I think there's still something a bit funky here, but most of this gets deleted in the next PR so, let's just address what's left there

**Review by danhhz — APPROVED** [review]
🚢

**Comment by ruchirK** [comment]
tftr! merging on green!
